cli/ssh_test.go

```diff
@@ -86,6 +91,62 @@ func TestSSH(t *testing.T) {
 		pty.WriteLine("exit")
 		<-cmdDone
 	})
+	t.Run("StartStoppedWorkspace", func(t *testing.T) {
+		t.Parallel()
+
+		//store, ps := dbtestutil.NewDB(t)
+		//ownerClient := coderdtest.New(t, &coderdtest.Options{Pubsub: ps, Database: store, IncludeProvisionerDaemon: true})
+		//ownerClient.SetLogger(slogtest.Make(t, nil).Named("client").Leveled(slog.LevelDebug))
+		//first := coderdtest.CreateFirstUser(t, ownerClient)
+		//client, user := coderdtest.CreateAnotherUser(t, ownerClient, first.OrganizationID)
+		//
+		//// Create a stopped workspace.
+		//workspaceData := dbfake.WorkspaceBuild(t, store, database.Workspace{
+		//	OrganizationID: first.OrganizationID,
+		//	OwnerID:        user.ID,
+		//}).Seed(database.WorkspaceBuild{
+		//	Transition: database.WorkspaceTransitionStop,
+		//}).WithAgent().Do()
+		//workspace := workspaceData.Workspace
+		//
+		authToken := uuid.NewString()
+		ownerClient := coderdtest.New(t, &coderdtest.Options{IncludeProvisionerDaemon: true})
+		owner := coderdtest.CreateFirstUser(t, ownerClient)
+		client, _ := coderdtest.CreateAnotherUser(t, ownerClient, owner.OrganizationID, rbac.RoleTemplateAdmin())
+		version := coderdtest.CreateTemplateVersion(t, client, owner.OrganizationID, &echo.Responses{
+			Parse:          echo.ParseComplete,
+			ProvisionPlan:  echo.PlanComplete,
+			ProvisionApply: echo.ProvisionApplyWithAgent(authToken),
+		})
+		coderdtest.AwaitTemplateVersionJobCompleted(t, client, version.ID)
+		template := coderdtest.CreateTemplate(t, client, owner.OrganizationID, version.ID)
+		workspace := coderdtest.CreateWorkspace(t, client, owner.OrganizationID, template.ID)
+		coderdtest.AwaitWorkspaceBuildJobCompleted(t, client, workspace.LatestBuild.ID)
+		// Stop the workspace
+		workspaceBuild := coderdtest.CreateWorkspaceBuild(t, client, workspace, database.WorkspaceTransitionStop)
+		coderdtest.AwaitWorkspaceBuildJobCompleted(t, client, workspaceBuild.ID)
+
+		//
+		inv, root := clitest.New(t, "ssh", workspace.Name)
+		clitest.SetupConfig(t, client, root)
+		pty := ptytest.New(t).Attach(inv)
+
+		ctx, cancel := context.WithTimeout(context.Background(), testutil.WaitLong)
+		defer cancel()
+
+		cmdDone := tGo(t, func() {
+			err := inv.WithContext(ctx).Run()
+			assert.NoError(t, err)
+		})
+		pty.ExpectMatch("⧗ Running")
```

Comment:
I wonder if this has the potential to be racy, i.e. workspace started before we end up streaming logs. 🤔

Reply:
That is a good point. I am unsure how to prevent that :thinking:.

We really should have a controllable provisioner in our unit tests to trigger this stuff :cry:

Reply:
Actually, we have `coderdtest.AwaitWorkspaceAgents(t, client, workspace.ID)` before I run `exit` which is sufficient to wait for the workspace to be ready.

Reply:
Yeah, I think workspace being ready shouldn't be an issue. But I'm wondering if the provisioner log missing could be an issue depending on at what stage of the workspace lifecycle we connect to it.

Reply:
Yup, I removed it because I check if the agent is ready right below it. So That agent ready will always work, even if we miss the logs.